pkg/block/fetcher.go

```diff
@@ -797,13 +818,23 @@ func NewIgnoreDeletionMarkFilter(logger log.Logger, bkt objstore.InstrumentedBuc
 
 // DeletionMarkBlocks returns block ids that were marked for deletion.
 func (f *IgnoreDeletionMarkFilter) DeletionMarkBlocks() map[ulid.ULID]*metadata.DeletionMark {
-	return f.deletionMarkMap
+	f.mtx.Lock()
+	defer f.mtx.Unlock()
+
+	deletionMarkMap := make(map[ulid.ULID]*metadata.DeletionMark)
+	for id, meta := range f.deletionMarkMap {
+		deletionMarkMap[id] = meta
+	}
+
+	return deletionMarkMap
 }
 
 // Filter filters out blocks that are marked for deletion after a given delay.
 // It also returns the blocks that can be deleted since they were uploaded delay duration before current time.
 func (f *IgnoreDeletionMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced *extprom.TxGaugeVec) error {
+	f.mtx.Lock()
 	f.deletionMarkMap = make(map[ulid.ULID]*metadata.DeletionMark)
```

Comment:
Why is that needed, do we access this map somewhere else?

If that's the case I think we have to build this map locally and only then swap in locked fashion, right?

Reply:
Otherwise we will have "partial" results

Reply:
It's being accessed in [`DeletionMarkBlocks`](DeletionMarkBlocks). I'm wondering what are the ramifications of having only 'partial' results in the map, shouldn't we pick up any leftovers during the next run of the cleaner func? (Isn't that how it works now, as current version is also potentially allowing to access map with 'partial' results?).

Otherwise you're right, we can also build it separately and swap afterwards while using lock.

Reply:
I am just assuming there is a reader somewhere else accessing the deletion map concurrently. For a short period instead of 20000 deleted items it would have zero elements and the back to 200 000. I would need to check what exactly wrong can happen with this, but it is just asking for problems here, no? (: Maybe now, things would eventually heal, but tomorrow no. No one would assume that internal map is flaky unless it's called `flakyDeleionMarkMap` across instances with 20 lines of comments on why and still that might be surprising for next writer/reader (:

Reply:
I would fix this before moving on

Reply:
Thanks for clarifying, I see the point now :+1: I adjusted it, please let me know if it looks good now!